src/t_hash.c

```diff
@@ -434,25 +1116,65 @@ void hashTypeConvertListpack(robj *o, int enc) {
     if (enc == OBJ_ENCODING_LISTPACK) {
         /* Nothing to do... */
 
+    } else if (enc == OBJ_ENCODING_LISTPACK_TTL) {
+        void *fptr;
+        long long val;
+        unsigned int slen;
+
+        /* We are going to store HASH_LP_NO_TTL (zero) as the TTL. 'zero'
+         * occupies two bytes in the listpack. Estimate space for additional
+         * 2 bytes per 'pair': Number of pairs is 'lpLength(o->ptr) / 2'
+         * Multiply it with 2 bytes: lpLength() / 2 * 2 = lpLength() */
+        size_t cap = lpBytes(o->ptr) + lpLength(o->ptr);
+        listpackTTL *lpt = listpackTTLCreate(cap);
+
+        fptr = lpFirst(o->ptr);
+        while (fptr != NULL) {
+            void *ptr;
+
+            ptr = lpGetValue(fptr, &slen, &val);
+            if (ptr)
+                lpt->lp = lpAppend(lpt->lp, ptr, slen);
+            else
+                lpt->lp = lpAppendInteger(lpt->lp, val);
+
+            fptr = lpNext(o->ptr, fptr);
+            serverAssert(fptr);
+
+            ptr = lpGetValue(fptr, &slen, &val);
+            if (ptr)
+                lpt->lp = lpAppend(lpt->lp, ptr, slen);
+            else
+                lpt->lp = lpAppendInteger(lpt->lp, val);
+
+            lpt->lp = lpAppendInteger(lpt->lp, HASH_LP_NO_TTL);
+            fptr = lpNext(o->ptr, fptr);
+        }
```

Comment:
Sun: not sure if just using lpInsertInteger(.., HASH_LP_NO_TTL, ...) on the old lp would be faster.

Reply:
done, it is faster, for most cases